networkx/algorithms/operators/binary.py

```diff
@@ -332,6 +333,87 @@ def compose(G, H):
+    For a multigraph, given m edges from a to b and n edges from b to c, the result will have mxn edges from a to c.  The user can define attributes of these edges as a function edge_data_combiner of the attributes of a participating edge instance for each particpating edge.
+    Where two nodes are connected by more than one intermediate node, a single edge is created in a Graph or DiGraph (it is not defined in this case which pair is used to calculate the attribute data), and multiple edges are created in a MultiGraph or MultiDiGraph (so attribute data calculated from each pair is preserved), but note the restrictions imposed by the with_keys parameter.
+    An undirected graph is treated as a directed graph with dual directed edges between the connected nodes, in each direction, as by to_directed, so can be composed with a directed graph in either direction.
+    The relational composition of hypergraphs would contain a hyperedge with nodeset U if G contains a hyperedge with nodeset V and H contains a hyperedge with nodeset W, with U the symmetric difference of V and W.
+
+    Parameters
+    ----------
+    G, H : graph
+       A NetworkX graph
+    with_keys : bool, optional (default=False)
+       Ignored for simple graphs.
+       For multigraphs:
+          If with_keys=True, keys of edges being composed must match.  This may be desired if keys are set explicitly.
+          Otherwise, keys are ignored and reassigned.
+    with_data : bool, optional (default=True)
+    edge_data_combiner: function of two arguments
+
+    Returns
+    -------
+    C: A new graph that is a multigraph if both G and H are multigraphs; directed if either G or H is directed; an undirected graph if both G and H are undirected graphs; and an error otherwise
+
+    Notes
+    -----
+    Attributes from the graph and nodes are copied to the new graph if with_data is set.  For consistency with other routines, attributes from H take precedence over attributes from G.
+    Attributes from the edges are copied to the new graph if with_data is set and a edge_data_combiner is provided.
+    """
+    # specify result graph
+    if not (G.is_multigraph() or H.is_multigraph()):
+        is_multigraph = False
+    else:
+        if not G.is_multigraph():
+            G = G.to_multigraph()
+        elif not H.is_multigraph():
+            H = H.to_multigraph()
+        is_multigraph = True
+
+    if not (G.is_directed() or H.is_directed()):
+        is_directed = False
+    else:
+        if not G.is_directed():
+            G = G.to_directed()
+        elif not H.is_directed():
+            H = H.to_directed()
+        is_directed = True
+
+    # create result graph
+    R = nx.create_empty_copy(G, with_data=with_data)
+    if with_data:
+        for k, v in H.graph.items():
+            R.graph[k] = v
+        R.add_nodes_from(H.nodes(data=True))
+
+    # add composite edges to result graph
+    for (gsource, nbrdict) in G.adjacency_iter():
```

Comment:
OK, thanks.  I see that G.adjacency_iter was replaced by G..adjacency, which is no longer supported.  I was referring to some old documentation and also there are still stray references to adjacency_iter in the current version.  No problem, I'll do it the way you suggest.

Reply:
Yeah, there used to be lots of different ways to do it. Since the the view overhaul there was work on consolidating things such that there was "one true way" to do everything. In this case the `.adj` attribute has everything you need.

Reply:
It's best to deprecate and then remove all traces of the previous ways.

Reply:
Where are the stray references to `adjacency_iter`?  We got rid of that years ago.

Reply:
Dan, You pointed out the definition of adjacency below, and there are various calls to it.  There's an adjacency_iter at examples/subclass/plot_antigraph.py which, although not directly applicable, led me to believe that defining such a function was the appropriate way to handle this.

Reply:
Thank you!  I can believe that we missed that one...  I'll open an issue.   :}